CrazyCanvas/Source/World/LevelObjectCreator.cpp

```diff
@@ -195,11 +200,74 @@ ESpecialObjectType LevelObjectCreator::CreateSpawnpoint(const LambdaEngine::Spec
-	UNREFERENCED_VARIABLE(createdEntities);
-	UNREFERENCED_VARIABLE(translation);
+	using namespace LambdaEngine;
+
+	if (specialObject.MeshComponents.GetSize() > 1)
+	{
+		LOG_WARNING("[LevelObjectCreator]: Special Object \"Flag\" has more than one Mesh Component, this is not currently supported");
+	}
+
+	const MeshComponent& meshComponent = specialObject.MeshComponents[0];
+
+	ECSCore* pECS					= ECSCore::GetInstance();
+	PhysicsSystem* pPhysicsSystem	= PhysicsSystem::GetInstance();
+
+	Entity entity = pECS->CreateEntity();
+
+	pECS->AddComponent<OffsetComponent>(entity, OffsetComponent{ .Offset = glm::vec3(0.0f)});
+	pECS->AddComponent<ParentComponent>(entity, ParentComponent{ .Attached = false });
+	const CollisionCreateInfo collisionCreateInfo =
+	{
+		.Entity				= entity,
+		.Position			= pECS->AddComponent<PositionComponent>(entity,		{ true, specialObject.DefaultPosition/* + translation + glm::vec3(0.0f, specialObject.DefaultScale.y * specialObject.BoundingBoxes[0].HalfExtent.y, 0.0f)*/}),
+		.Scale				= pECS->AddComponent<ScaleComponent>(entity,		{ true, specialObject.DefaultScale }),
+		.Rotation			= pECS->AddComponent<RotationComponent>(entity,		{ true, specialObject.DefaultRotation }),
+		.Mesh				= pECS->AddComponent<MeshComponent>(entity,			meshComponent),
+		.ShapeType			= EShapeType::TRIGGER,
+		.CollisionGroup		= FCrazyCanvasCollisionGroup::COLLISION_GROUP_FLAG,
+		.CollisionMask		= FCrazyCanvasCollisionGroup::COLLISION_GROUP_PLAYER, // Collide with any non-static object
+		.CollisionCallback	= [](const EntityCollisionInfo& collisionInfo0, const EntityCollisionInfo& collisionInfo1)
+		{
+			ECSCore* pECS = ECSCore::GetInstance();
+
+			LOG_WARNING("FLAG COLLISION");
+
+			Entity flagEntity	= collisionInfo0.Entity;
+			Entity playerEntity	= collisionInfo1.Entity;
+
+			const MeshComponent& playerMeshComponent			= pECS->GetConstComponent<MeshComponent>(playerEntity);
+
+			StaticCollisionComponent& flagCollisionComponent	= pECS->GetComponent<StaticCollisionComponent>(flagEntity);
+			ParentComponent& flagParentComponent				= pECS->GetComponent<ParentComponent>(flagEntity);
+			OffsetComponent& flagOffsetComponent				= pECS->GetComponent<OffsetComponent>(flagEntity);
+
+			PxShape* pFlagShape;
+			flagCollisionComponent.pActor->getShapes(&pFlagShape, 1);
+			
+			//Update Collision Group
+			PxFilterData filterData;
+			filterData.word0 = (PxU32)FCrazyCanvasCollisionGroup::COLLISION_GROUP_FLAG;
+			filterData.word1 = (PxU32)FCollisionGroup::COLLISION_GROUP_NONE;
+			pFlagShape->setSimulationFilterData(filterData);
+			pFlagShape->setQueryFilterData(filterData);
+
+			//Set Flag Carrier (Parent)
+			flagParentComponent.Attached	= true;
+			flagParentComponent.Parent		= playerEntity;
+
+			//Set Flag Offset
+			const Mesh* pMesh = ResourceManager::GetMesh(playerMeshComponent.MeshGUID);
+			flagOffsetComponent.Offset		= glm::vec3(0.0f, pMesh->BoundingBox.Dimensions.y / 2.0f, 0.0f);
+
+			Level::OnFlagPickedUp();
+		}
```

Comment:
Two things about this callback function:
#### Where to place the callback function definition
I think the definition of the callback function should be somewhere else. I think it's best to have it as a member function of some class handling gameplay logic.

#### The callback function should schedule a job
The callback function makes accesses to components when other systems and jobs might be running. To avoid data races and other threading errors, such accesses should be made within a job. A job consists of two things:
1. An array of component types accessed during the job's execution, and with what permissions (read or read-write).
2. A function to execute

TLDR; create a job within the callback function, and schedule it using `ECSCore::ScheduleJobASAP(...) `or `ECSCore::ScheduleJobPostFrame(...)`. The latter executes after all systems have finished.

Reply:
The job thing I agree with, but the callback function placement I don't really agree with.

My reasoning behind this is that I would like `LevelObjectCreator ` to be the main place where we define all Level Entity functionality, mainly to avoid jumping between so many different files when changing stuff, which we have to do for some other things right now and I find it annoying.